drivers/sensor/bosch/bmm350/bmm350.c

```diff
@@ -0,0 +1,984 @@
+	uint8_t rx_buf[3] = {0x00};
+
+	if (pmu_cmd_stat_0 != NULL) {
+		/* Get PMU command status 0 data */
+		ret = bmm350_reg_read(dev, BMM350_REG_PMU_CMD_STATUS_0, &rx_buf[0], 3);
+		LOG_DBG("pmu cmd status 0:0x%x\n", rx_buf[2]);
+		if (ret == BMM350_OK) {
+			pmu_cmd_stat_0->pmu_cmd_busy =
+				BMM350_GET_BITS_POS_0(rx_buf[2], BMM350_PMU_CMD_BUSY);
+			pmu_cmd_stat_0->odr_ovwr = BMM350_GET_BITS(rx_buf[2], BMM350_ODR_OVWR);
+			pmu_cmd_stat_0->avr_ovwr = BMM350_GET_BITS(rx_buf[2], BMM350_AVG_OVWR);
+			pmu_cmd_stat_0->pwr_mode_is_normal =
+				BMM350_GET_BITS(rx_buf[2], BMM350_PWR_MODE_IS_NORMAL);
+			pmu_cmd_stat_0->cmd_is_illegal =
+				BMM350_GET_BITS(rx_buf[2], BMM350_CMD_IS_ILLEGAL);
+			pmu_cmd_stat_0->pmu_cmd_value =
+				BMM350_GET_BITS(rx_buf[2], BMM350_PMU_CMD_VALUE);
+		}
+	} else {
+		ret = BMM350_E_NULL_PTR;
+	}
+	return ret;
+}
+
+/*!
+ * @brief This internal API is used to switch from suspend mode to normal mode or forced mode.
+ */
+static int8_t set_powermode(enum bmm350_power_modes powermode, const struct device *dev)
+{
+	/* Variable to store the function result */
+	int8_t ret = 0;
+	uint8_t rx_buf[3] = {0x00};
+	uint8_t reg_data = powermode;
+
+	/* Array to store suspend to forced mode delay */
+	uint32_t sus_to_forced_mode[4] = {
+		BMM350_SUS_TO_FORCEDMODE_NO_AVG_DELAY, BMM350_SUS_TO_FORCEDMODE_AVG_2_DELAY,
+		BMM350_SUS_TO_FORCEDMODE_AVG_4_DELAY, BMM350_SUS_TO_FORCEDMODE_AVG_8_DELAY};
+
+	/* Array to store suspend to forced mode fast delay */
+	uint32_t sus_to_forced_mode_fast[4] = {BMM350_SUS_TO_FORCEDMODE_FAST_NO_AVG_DELAY,
+					       BMM350_SUS_TO_FORCEDMODE_FAST_AVG_2_DELAY,
+					       BMM350_SUS_TO_FORCEDMODE_FAST_AVG_4_DELAY,
+					       BMM350_SUS_TO_FORCEDMODE_FAST_AVG_8_DELAY};
+
+	uint8_t avg = 0;
+	uint32_t delay_us = 0;
+
+	if (ret == BMM350_OK) {
+		/* Get average configuration */
+		ret = bmm350_reg_read(dev, BMM350_REG_PMU_CMD_AGGR_SET, &rx_buf[0], 3);
+		/* Mask the average value */
+		avg = ((rx_buf[2] & BMM350_AVG_MSK) >> BMM350_AVG_POS);
+		if (ret == BMM350_OK) {
+			/* Get average configuration */
+			if (powermode == BMM350_NORMAL_MODE) {
+				delay_us = BMM350_SUSPEND_TO_NORMAL_DELAY;
+			}
+
+			/* Check if desired power mode is forced mode */
+			if (powermode == BMM350_FORCED_MODE) {
+				/* Store delay based on averaging mode */
+				delay_us = sus_to_forced_mode[avg];
+			}
+
+			/* Check if desired power mode is forced mode fast */
+			if (powermode == BMM350_FORCED_MODE_FAST) {
+				/* Store delay based on averaging mode */
+				delay_us = sus_to_forced_mode_fast[avg];
+			}
+
+			/* Set PMU command configuration to desired power mode */
+			ret = bmm350_reg_write(dev, BMM350_REG_PMU_CMD, reg_data);
+			k_usleep(delay_us);
+		}
+	}
+
+	LOG_DBG("pmu cmd agget set powermode %d\n", powermode);
+
+	return ret;
+}
+
+/*!
+ * @brief This API is used to set the power mode of the sensor
+ */
+static int8_t bmm350_set_powermode(enum bmm350_power_modes powermode, const struct device *dev)
```

Comment:
nit: usually, I always see the `const struct device *` as the first arg for every function. Just something that seems to not follow a pattern